xhttp/retriable/retriable.go

```diff
@@ -0,0 +1,477 @@
+
+// lenReader is an interface implemented by many in-memory io.Reader's. Used
+// for automatically sending the right Content-Length header when possible.
+type lenReader interface {
+	Len() int
+}
+
+//Requestor defines interface to make HTTP calls
+type Requestor interface {
+	Do(r *http.Request) (*http.Response, error)
+}
+
+// ReaderFunc is the type of function that can be given natively to NewRequest
+type ReaderFunc func() (io.Reader, error)
+
+// Request wraps the metadata needed to create HTTP requests.
+type Request struct {
+	// body is a seekable reader over the request body payload. This is
+	// used to rewind the request data in between retries.
+	body ReaderFunc
+
+	// Embed an HTTP request directly. This makes a *Request act exactly
+	// like an *http.Request so that all meta methods are supported.
+	*http.Request
+}
+
+// NewRequest creates a new wrapped request.
+func NewRequest(method, url string, rawBody io.ReadSeeker) (*Request, error) {
+	var body ReaderFunc
+	var contentLength int64
+
+	if rawBody != nil {
+		raw := rawBody.(io.ReadSeeker)
+		body = func() (io.Reader, error) {
+			raw.Seek(0, 0)
+			return ioutil.NopCloser(raw), nil
+		}
+		if lr, ok := raw.(lenReader); ok {
+			contentLength = int64(lr.Len())
+		}
+	}
+
+	httpReq, err := http.NewRequest(method, url, nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	httpReq.ContentLength = contentLength
+
+	return &Request{body, httpReq}, nil
+}
+
+// convertRequest wraps http.Request into retry.Request
+func convertRequest(req *http.Request) (*Request, error) {
+	var body io.ReadSeeker
+	if req != nil && req.Body != nil {
+		defer req.Body.Close()
+		bodyBytes, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		body = bytes.NewReader(bodyBytes)
+	}
+
+	r, err := NewRequest(req.Method, req.URL.String(), body)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	r.Request = r.WithContext(req.Context())
+	for header, vals := range req.Header {
+		for _, val := range vals {
+			r.Request.Header.Add(header, val)
+		}
+	}
+
+	return r, nil
+}
+
+// ShouldRetry specifies a policy for handling retries. It is called
+// following each request with the response, error values returned by
+// the http.Client and the number of already made retries.
+// If ShouldRetry returns false, the Client stops retrying
+// and returns the response to the caller. The
+// Client will close any response body when retrying, but if the retry is
+// aborted it is up to the caller to properly close any response body before returning.
+type ShouldRetry func(resp *http.Response, err error, retries int) (bool, time.Duration, string)
+
+// Policy represents the retry policy
+type Policy struct {
+
+	// Retries specifies a map of HTTP Status code to ShouldRetry function,
+	// 0 status code indicates a connection related error (network, TLS, DNS etc.)
+	Retries map[int]ShouldRetry
```

Comment:
@haykbaluyan I've refactored your `retry` package to `xhttp/retriable`.
Can you please look at the tests file, and port old tests, but make them generic, without specific REST api?